source/common/buffer/watermark_buffer.cc

```diff
@@ -136,5 +140,101 @@ void WatermarkBuffer::checkHighAndOverflowWatermarks() {
   }
 }
 
+BufferMemoryAccountSharedPtr
+WatermarkBufferFactory::createAccount(Http::StreamResetHandler* reset_handler) {
+  return BufferMemoryAccountImpl::createAccount(this, reset_handler);
+}
+
+void WatermarkBufferFactory::updateAccountClass(const BufferMemoryAccountSharedPtr& account,
+                                                int current_class, int new_class) {
+  ASSERT(current_class != new_class, "Expected the current_class and new_class to be different");
+
+  if (current_class == -1 && new_class >= 0) {
+    // Start tracking
+    ASSERT(!size_class_account_sets_[new_class].contains(account));
+    size_class_account_sets_[new_class].insert(account);
+  } else if (current_class >= 0 && new_class == -1) {
+    // No longer track
+    ASSERT(size_class_account_sets_[current_class].contains(account));
+    size_class_account_sets_[current_class].erase(account);
+  } else {
+    // Moving between buckets
+    ASSERT(size_class_account_sets_[current_class].contains(account));
+    ASSERT(!size_class_account_sets_[new_class].contains(account));
+    size_class_account_sets_[new_class].insert(
+        std::move(size_class_account_sets_[current_class].extract(account).value()));
+  }
+}
+
+void WatermarkBufferFactory::unregisterAccount(const BufferMemoryAccountSharedPtr& account,
+                                               int current_class) {
+  if (current_class >= 0) {
+    ASSERT(size_class_account_sets_[current_class].contains(account));
+    size_class_account_sets_[current_class].erase(account);
+  }
+}
+
+WatermarkBufferFactory::~WatermarkBufferFactory() {
+  for (auto& account_set : size_class_account_sets_) {
+    ASSERT(account_set.empty(),
+           "Expected all Accounts to have unregistered from the Watermark Factory.");
+  }
+}
+
+BufferMemoryAccountSharedPtr
+BufferMemoryAccountImpl::createAccount(WatermarkBufferFactory* factory,
+                                       Http::StreamResetHandler* reset_handler) {
+  // We use shared_ptr ctor directly rather than make shared since the
+  // constructor being invoked is private as we want users to use this static
+  // method to createAccounts.
+  auto account =
+      std::shared_ptr<BufferMemoryAccount>(new BufferMemoryAccountImpl(factory, reset_handler));
+  // Set shared_this_ in the account.
+  static_cast<BufferMemoryAccountImpl*>(account.get())->shared_this_ = account;
+  return account;
+}
+
+int BufferMemoryAccountImpl::balanceToClassIndex() {
+  uint64_t shifted_balance = buffer_memory_allocated_ >> 20; // shift by 1MB.
+
+  if (shifted_balance == 0) {
+    return -1; // Not worth tracking anything < 1MB.
```

Comment:
You're not allowing for a lot of configuration for this accounting.  Will we need some configuration in the future?

Note that the recommended stream buffer limits for edge configurations are around 64kb.  Tracking high memory connections at 1MB or larger leaves a fairly wide gap of ~16x before this mechanism can terminate expensive streams.  It seems very unlikely that we'll have streams using >1MB given the 64kb edge config stream limit.

Reply:
I would consider requiring that the config be powers of two.  I do think that the addition of config will force you to undo some of the micro optimizations in this component.

Reply:
Implemented this in https://github.com/KBaichoo/envoy/pull/105/